### PR TITLE
GH-3690: Support customization of RedisVectorStore via builder customizer

### LINKED ISSUE #3690: RedisVectorStore auto-configuration does not support metadataFields, causing chat memory retrieval failure
**Bug description**  
When using `VectorStoreChatMemoryAdvisor` to implement chat memory with Redis vector store, if `spring.ai.vectorstore.redis.initialize-schema` is set to `true`, the program will automatically create a Redis search schema, which causes `RedisVectorStore` to be unable to retrieve related records.

The reason is that `VectorStoreChatMemoryAdvisor` writes `"conversationId"` as metadata into the document and uses `conversationId` as the query condition. However, `RedisVectorStoreProperties` does not support setting `metadataFields` when automatically configuring `RedisVectorStore`, leading to retrieval failure.

It is necessary to manually configure `RedisVectorStore` to retrieve records correctly, like this (see code below), which is not elegant because it defeats the purpose of auto-configuration.

```java
@Bean
public RedisVectorStore vectorStore(EmbeddingModel embeddingModel, RedisVectorStoreProperties properties,
                                    JedisConnectionFactory jedisConnectionFactory, ObjectProvider<ObservationRegistry> observationRegistry,
                                    ObjectProvider<VectorStoreObservationConvention> customObservationConvention,
                                    BatchingStrategy batchingStrategy) {

    JedisPooled jedisPooled = this.jedisPooled(jedisConnectionFactory);
    return RedisVectorStore.builder(jedisPooled, embeddingModel)
            .initializeSchema(properties.isInitializeSchema())
            .observationRegistry(observationRegistry.getIfUnique(() -> ObservationRegistry.NOOP))
            .customObservationConvention(customObservationConvention.getIfAvailable(() -> null))
            .batchingStrategy(batchingStrategy)
            .indexName(properties.getIndexName())
            .prefix(properties.getPrefix())
            // Set metadata fields to include conversationId for retrieval
            .metadataFields(RedisVectorStore.MetadataField.tag("conversationId"))
            .build();
}
```

**Environment**
- Spring Boot: 3.4.6
- Spring AI: 1.0.0
- Java: 21
- Vector store: redis-stack v7.2.0

**Steps to reproduce**
1. Use `VectorStoreChatMemoryAdvisor` with Redis vector store.
```xml
<dependency>
    <groupId>org.springframework.ai</groupId>
    <artifactId>spring-ai-advisors-vector-store</artifactId>
    <version>1.0.0</version>
</dependency>
<dependency>
    <groupId>org.springframework.ai</groupId>
    <artifactId>spring-ai-starter-vector-store-redis</artifactId>
    <version>1.0.0</version>
</dependency>
```
```java
// configuration
@Bean
public ChatClient chatClient(ChatClient.Builder clientBuilder, VectorStore vectorStore) {
    return clientBuilder
            .defaultSystem("You are a helpful assistant.")
            .defaultAdvisors(VectorStoreChatMemoryAdvisor.builder(vectorStore).build())
            .build();
}
```
2. Set `spring.ai.vectorstore.redis.initialize-schema=true`.
```yaml
spring:
  data:
    redis:
      client-type: jedis
  ai:
    ollama:
      chat:
        options:
          model: deepseek-r1
      embedding:
        options:
          model: bge-m3
    vectorstore:
      redis:
        initialize-schema: true
```
3. Attempt to retrieve chat memory by `conversationId` in a conversation.
```java
// controller
@GetMapping(value = "/stream/chat", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
public Flux<String> streamChat(@RequestParam @NotBlank String prompt, @RequestParam @NotBlank String conversationId) {
    return chatClient
            .prompt(Prompt.builder().content(prompt).build())
            .advisors(advisorSpec -> advisorSpec.param(ChatMemory.CONVERSATION_ID, conversationId))
            .stream().content();
}
```

**Expected behavior**  
Chat memory relevant to the current `conversationId` should be retrievable automatically without manual configuration of `RedisVectorStore`.

**Possible solutions**
1. Set `spring.ai.vectorstore.redis.initialize-schema=false` and manually create the schema in Redis.
2. Add a `List<MetadataField> metadataFields` property to `RedisVectorStoreProperties`.
3. Use `RedisVectorStore.Builder` as a dependency to build `RedisVectorStore`, where `RedisVectorStore.Builder` depends on `RedisVectorStoreProperties` and provides a `RedisVectorStoreBuilderCustomizer` for builder enhancement, like this:
```java
// VectorStoreBuilderCustomizer.java
public interface VectorStoreBuilderCustomizer<T extends VectorStore.Builder<T>> {
    void customize(T builder);
}
// RedisVectorStoreBuilderCustomizer.java
public interface RedisVectorStoreBuilderCustomizer extends VectorStoreBuilderCustomizer<RedisVectorStore.Builder> {

}
// RedisVectorStoreAutoConfiguration.java
@Bean
@ConditionalOnMissingBean
public RedisVectorStore.Builder vectorStoreBuilder(EmbeddingModel embeddingModel, RedisVectorStoreProperties properties,
                                    JedisConnectionFactory jedisConnectionFactory, ObjectProvider<ObservationRegistry> observationRegistry,
                                    ObjectProvider<VectorStoreObservationConvention> customObservationConvention,
                                    BatchingStrategy batchingStrategy, ObjectProvider<RedisVectorStoreBuilderCustomizer> vectorStoreBuilderCustomizers) {

    JedisPooled jedisPooled = this.jedisPooled(jedisConnectionFactory);
    RedisVectorStore.Builder builder = RedisVectorStore.builder(jedisPooled, embeddingModel)
            .initializeSchema(properties.isInitializeSchema())
            .observationRegistry(observationRegistry.getIfUnique(() -> ObservationRegistry.NOOP))
            .customObservationConvention(customObservationConvention.getIfAvailable(() -> null))
            .batchingStrategy(batchingStrategy)
            .indexName(properties.getIndexName())
            .prefix(properties.getPrefix());
    vectorStoreBuilderCustomizers.orderedStream().forEach(customizer -> customizer.customize(builder));
    return builder;
}

@Bean
@ConditionalOnMissingBean
public RedisVectorStore vectorStore(RedisVectorStore.Builder vectorStoreBuilder) {
    return vectorStoreBuilder.build();
}
```

I am not sure why `RedisVectorStoreProperties` does not directly provide a `metadataFields` property. If there is a reason, please consider solution 3. In fact, solution 3 does not explicitly depend on `RedisVectorStoreProperties`, but instead uses `RedisVectorStore.Builder` to determine how to build the `RedisVectorStore`.

If you are open to my suggestion, I can propose a PR.

### PR DESCRIPTION
**Pull request by Acacian**
GH-3690: Add RedisVectorStore builder customization support

Closes #3690

This PR introduces an extension point to customize `RedisVectorStore.Builder` via `RedisVectorStoreBuilderCustomizer`.

---

## 📌 Background

When using `VectorStoreChatMemoryAdvisor` with RedisVectorStore and `initialize-schema=true`,  
metadata fields such as `conversationId` (used in `filterExpression`) are not registered in the Redis schema,  
leading to **retrieval failures** during similarity search.

This issue was raised by @lanpf in #3690, and we experimentally adopted the direction they proposed —  
allowing schema-related settings to be injected through a `Customizer`, without overriding auto-configuration.

---

## ✨ Changes

- Add a generic extension point:  
  `VectorStoreBuilderCustomizer<T extends VectorStore.Builder<T>>`
- Specialize it for Redis:  
  `RedisVectorStoreBuilderCustomizer extends VectorStoreBuilderCustomizer<RedisVectorStore.Builder>`
- Apply ordered customizers in `RedisVectorStoreAutoConfiguration`:

```java
customizers.orderedStream().forEach(c -> c.customize(builder));
```

This provides a **Spring-style Customizer Pattern**, enabling open extension without modifying the default auto-configuration.

---

## ✅ Example

Developers can now customize metadata fields non-invasively like this:

```java
@Bean
public RedisVectorStoreBuilderCustomizer metadataCustomizer() {
    return builder -> builder.metadataFields(List.of(RedisVectorStore.MetadataField.tag("conversationId")));
}
```

This allows schema initialization and metadata filter-based retrieval to work together as expected —  
without disabling `initialize-schema=true` or manually wiring a `RedisVectorStore`.

---

## 🧪 Test & Compliance

- ✔️ `./mvnw clean verify` all passed  
- ✔️ Checkstyle passed  
- ✔️ DCO signed: `Signed-off-by: Dongha Koo <koo.dongha@gmail.com>`  
- ✔️ Rebased on latest `main`, linear commit history preserved

We hope this solution addresses the needs expressed by @lanpf and others with similar issues.  
Looking forward to your review and feedback. Thank you!

### CODE REVIEW

**Comment by mengnankkkk** [comment]
This PR introduces a generic extension point for customizing RedisVectorStore.Builder via RedisVectorStoreBuilderCustomizer, addressing the issue where metadata fields (e.g., conversationId) are not registered in the Redis schema during auto-configuration, causing retrieval failures.
Possible improvements:
Currently only supports RedisVectorStore; consider generalizing to other VectorStore implementations.
Documentation could be expanded to clarify usage scenarios and edge cases.
Test coverage could be enhanced, e.g., for multiple customizers, error handling, and ordering.

**Comment by Acacian** [comment]
@mengnankkkk 

Thanks for the thoughtful review.

I’ll take your suggestions into consideration and plan to follow up with an additional commit soon to improve test coverage and documentation.

Appreciate your feedback!

File: vector-stores/spring-ai-redis-store/src/main/java/org/springframework/ai/vectorstore/redis/RedisVectorStoreBuilderCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package org.springframework.ai.vectorstore.redis;
+
+/**
+ * A customizer interface for RedisVectorStore builder.
+ *
+ * @author Dongha Koo
+ * @since 1.0.0
```
Comment (sunyuhan1998) [inline vector-stores/spring-ai-redis-store/src/main/java/org/springframework/ai/vectorstore/redis/RedisVectorStoreBuilderCustomizer.java]: Should this be 1.1.0 for the current version, or if that PR will be merged and backport to 1.0.x, we might want to consider just removing `@since`

Comment (Acacian) [reply vector-stores/spring-ai-redis-store/src/main/java/org/springframework/ai/vectorstore/redis/RedisVectorStoreBuilderCustomizer.java]: Thanks for the suggestion!  
Since this is my first contribution, I really appreciate the guidance.  
I'll update the PR to remove `@since` as advised.

Comment (Acacian) [reply vector-stores/spring-ai-redis-store/src/main/java/org/springframework/ai/vectorstore/redis/RedisVectorStoreBuilderCustomizer.java]: Resolved as suggested. Thanks again!

**Comment by Acacian** [comment]
📝 Update based on review comments:

- Removed `@since` tag as suggested by @sunyuhan1998 to avoid confusion regarding versioning and potential backporting.
- Addressed feedback from @mengnankkkk:
  - Generalized the customization mechanism via the new `VectorStoreBuilderCustomizer` interface
  - Improved documentation to clarify usage and customization scenarios
  - Added test coverage for single customizer behavior and metadata field injection

Thanks for the detailed suggestions!
Let me know if further adjustments are needed 🙏

**Comment by lanpf** [comment]
Thank you for your interest in this issue. However, I had already proposed a detailed solution with full code implementation in #3690, which closely matches your approach. Since this was my original work, proper attribution is required under open-source norms—this includes crediting me as the author and allowing me to submit the PR. These are fundamental expectations in open-source collaboration, and I trust you’ll respect these guidelines.
I would appreciate it if you could respect my original solution. @Acacian @mengnankkkk @sunyuhan1998 @rozza 

Sent from my iPhone

On Jul 12, 2025, at 13:37, 구동하 ***@***.***> wrote:

﻿
[https://avatars.githubusercontent.com/u/52133649?s=20&v=4]Acacian left a comment (spring-projects/spring-ai#3795)<https://github.com/spring-projects/spring-ai/pull/3795#issuecomment-3064708037>

📝 Update based on review comments:

  *   Removed @since tag as suggested by @sunyuhan1998<https://github.com/sunyuhan1998> to avoid confusion regarding versioning and potential backporting.
  *   Addressed feedback from @mengnankkkk<https://github.com/mengnankkkk>:
     *   Generalized the customization mechanism via the new VectorStoreBuilderCustomizer interface
     *   Improved documentation to clarify usage and customization scenarios
     *   Added test coverage for multiple customizers and metadata field configuration

Thanks for the detailed suggestions!
Let me know if further adjustments are needed 🙏

**Comment by lanpf** [comment]
I would like to formally request the opportunity to submit this PR, as it aligns with my original proposal in #3690 @mengnankkkk @Acacian @sunyuhan1998 @rozza

**Comment by Acacian** [comment]
Hi @lanpf,  
Thank you again for your detailed suggestion and for raising the issue in the first place — I really appreciate your input.

When I saw that the issue had been open for over two weeks with no PR submitted, and that the only maintainer action was labeling it as a "bug", I assumed the team was inviting community contributions. Your comment suggested you were considering a PR, but since there was no follow-up, I proceeded with a fix — carefully referencing your proposal in the PR description and aligning with the contribution guidelines.

I also made sure that the contribution aligned with the official contribution guidelines, so there shouldn’t be any concerns from a process standpoint.

I referenced your suggestion explicitly in the PR because I genuinely wanted to acknowledge your input and show that your proposal was taken seriously. That recognition was important to me as part of contributing respectfully to the project.

**Comment by lanpf** [comment]
> Hi @lanpf, Thank you again for your detailed suggestion and for raising the issue in the first place — I really appreciate your input.
> 
> When I saw that the issue had been open for over two weeks with no PR submitted, and that the only maintainer action was labeling it as a "bug", I assumed the team was inviting community contributions. Your comment suggested you were considering a PR, but since there was no follow-up, I proceeded with a fix — carefully referencing your proposal in the PR description and aligning with the contribution guidelines.
> 
> I also made sure that the contribution aligned with the official contribution guidelines, so there shouldn’t be any concerns from a process standpoint.
> 
> I referenced your suggestion explicitly in the PR because I genuinely wanted to acknowledge your input and show that your proposal was taken seriously. That recognition was important to me as part of contributing respectfully to the project.


If your final solution differed from mine, I would have no issue with you resolving this bug. However, you clearly noticed my proposed code and solution in the issue, ultimately using them without privately consulting me for permission.

I shared my fix and detailed approach in the issue specifically to help others facing the same problem (as evidenced by the user who encountered it and my reply suggesting they use my code temporarily while awaiting community feedback).

This isn’t just about ‘finding and fixing a bug’—it’s a matter of originality and, arguably, copyright. If the community condones this behavior, it would violate the spirit of open-source by incentivizing contributors to withhold code prematurely, even when others need assistance.

**Comment by lanpf** [comment]
> Hi @lanpf, Thank you again for your detailed suggestion and for raising the issue in the first place — I really appreciate your input.
> 
> When I saw that the issue had been open for over two weeks with no PR submitted, and that the only maintainer action was labeling it as a "bug", I assumed the team was inviting community contributions. Your comment suggested you were considering a PR, but since there was no follow-up, I proceeded with a fix — carefully referencing your proposal in the PR description and aligning with the contribution guidelines.
> 
> I also made sure that the contribution aligned with the official contribution guidelines, so there shouldn’t be any concerns from a process standpoint.
> 
> I referenced your suggestion explicitly in the PR because I genuinely wanted to acknowledge your input and show that your proposal was taken seriously. That recognition was important to me as part of contributing respectfully to the project.

What's worse - this essentially encourages plagiarism. I shared my solution openly to help others facing the same issue, not to have it taken without proper attribution. When community members reproduce others' work without acknowledgment, it creates a toxic environment where contributors become reluctant to share their solutions early. This undermines the very purpose of open-source collaboration.

**Comment by Acacian** [comment]
> > Hi @lanpf, Thank you again for your detailed suggestion and for raising the issue in the first place — I really appreciate your input.
> > When I saw that the issue had been open for over two weeks with no PR submitted, and that the only maintainer action was labeling it as a "bug", I assumed the team was inviting community contributions. Your comment suggested you were considering a PR, but since there was no follow-up, I proceeded with a fix — carefully referencing your proposal in the PR description and aligning with the contribution guidelines.
> > I also made sure that the contribution aligned with the official contribution guidelines, so there shouldn’t be any concerns from a process standpoint.
> > I referenced your suggestion explicitly in the PR because I genuinely wanted to acknowledge your input and show that your proposal was taken seriously. That recognition was important to me as part of contributing respectfully to the project.
> 
> What's worse - this essentially encourages plagiarism. I shared my solution openly to help others facing the same issue, not to have it taken without proper attribution. When community members reproduce others' work without acknowledgment, it creates a toxic environment where contributors become reluctant to share their solutions early. This undermines the very purpose of open-source collaboration.

I understand the concern, and I want to emphasize again that it was never my intention to take credit for your original idea.

In fact, I made sure to reference your suggestion directly in the PR description to acknowledge your input and to show that it was taken seriously. That attribution was intentional and done out of respect.

I agree that open-source thrives when contributions are recognized and shared in good faith. At the same time, the model also encourages collaboration, where ideas proposed publicly are available for the community to act on — especially when they haven’t yet been formalized in a PR.

This PR was created in that spirit — to help move the project forward — while recognizing the value of your contribution.

If there are ways to improve attribution further, I’m happy to work with the maintainers on that. My only goal is to contribute constructively and respectfully.

**Comment by lanpf** [comment]
> > > Hi @lanpf, Thank you again for your detailed suggestion and for raising the issue in the first place — I really appreciate your input.
> > > When I saw that the issue had been open for over two weeks with no PR submitted, and that the only maintainer action was labeling it as a "bug", I assumed the team was inviting community contributions. Your comment suggested you were considering a PR, but since there was no follow-up, I proceeded with a fix — carefully referencing your proposal in the PR description and aligning with the contribution guidelines.
> > > I also made sure that the contribution aligned with the official contribution guidelines, so there shouldn’t be any concerns from a process standpoint.
> > > I referenced your suggestion explicitly in the PR because I genuinely wanted to acknowledge your input and show that your proposal was taken seriously. That recognition was important to me as part of contributing respectfully to the project.
> > 
> > 
> > What's worse - this essentially encourages plagiarism. I shared my solution openly to help others facing the same issue, not to have it taken without proper attribution. When community members reproduce others' work without acknowledgment, it creates a toxic environment where contributors become reluctant to share their solutions early. This undermines the very purpose of open-source collaboration.
> 
> I understand the concern, and I want to emphasize again that it was never my intention to take credit for your original idea.
> 
> In fact, I made sure to reference your suggestion directly in the PR description to acknowledge your input and to show that it was taken seriously. That attribution was intentional and done out of respect.
> 
> I agree that open-source thrives when contributions are recognized and shared in good faith. At the same time, the model also encourages collaboration, where ideas proposed publicly are available for the community to act on — especially when they haven’t yet been formalized in a PR.
> 
> This PR was created in that spirit — to help move the project forward — while recognizing the value of your contribution.
> 
> If there are ways to improve attribution further, I’m happy to work with the maintainers on that. My only goal is to contribute constructively and respectfully.

You could have handled this situation better through a more collaborative approach - perhaps by sending me a private message or commenting on my issue to discuss your intentions. Instead, you proceeded without any communication, and I don't even see my name credited in the authorship. While I understand you wanted to move the project forward, I believe there was ample time before the next release. The code and solution I publicly shared in my issue were sufficient for the interim period, which was precisely why I disclosed them early.

I believe we need a proper resolution for this matter. Here's my proposal: I will submit the PR myself, but will ensure your contributions are properly acknowledged in the file's authorship section. This is the best way to respect original work and copyright. What are your thoughts on this solution?

**Comment by Acacian** [comment]
Thanks for your thoughts. I truly understand and respect your passion for the code — as a fellow developer, I can relate to the sense of ownership and care that comes from crafting a good solution.

At the same time, I believe it's worth clarifying that open-source contributions are generally open to anyone, and PRs can be submitted by any contributor, especially when no PR has been created for more than two weeks after an issue is filed. That's why I assumed it was okay to proceed after referencing your suggestion in the PR description — it was never my intent to disregard your work.

That said, I appreciate your suggestion for a resolution. If you'd like to submit the PR yourself, that's completely fine with me. I’d be happy to close mine once yours is up. Just kindly mention me in the PR or commit to acknowledge the follow-up work, and we’re all good.

Let’s keep working together constructively — the ultimate goal is to make the project better.

**Comment by mengnankkkk** [comment]
> Thanks for your thoughts. I truly understand and respect your passion for the code — as a fellow developer, I can relate to the sense of ownership and care that comes from crafting a good solution.
> 
> At the same time, I believe it's worth clarifying that open-source contributions are generally open to anyone, and PRs can be submitted by any contributor, especially when no PR has been created for more than two weeks after an issue is filed. That's why I assumed it was okay to proceed after referencing your suggestion in the PR description — it was never my intent to disregard your work.
> 
> That said, I appreciate your suggestion for a resolution. If you'd like to submit the PR yourself, that's completely fine with me. I’d be happy to close mine once yours is up. Just kindly mention me in the PR or commit to acknowledge the follow-up work, and we’re all good.
> 
> Let’s keep working together constructively — the ultimate goal is to make the project better.




This is a small matter. We are all for the development of the project.Thanks for your contribution.

**Comment by lanpf** [comment]
> Thanks for your thoughts. I truly understand and respect your passion for the code — as a fellow developer, I can relate to the sense of ownership and care that comes from crafting a good solution.
> 
> At the same time, I believe it's worth clarifying that open-source contributions are generally open to anyone, and PRs can be submitted by any contributor, especially when no PR has been created for more than two weeks after an issue is filed. That's why I assumed it was okay to proceed after referencing your suggestion in the PR description — it was never my intent to disregard your work.
> 
> That said, I appreciate your suggestion for a resolution. If you'd like to submit the PR yourself, that's completely fine with me. I’d be happy to close mine once yours is up. Just kindly mention me in the PR or commit to acknowledge the follow-up work, and we’re all good.
> 
> Let’s keep working together constructively — the ultimate goal is to make the project better.

Thank you for your understanding. I'll submit the PR shortly and ensure your contributions to this issue are properly acknowledged.
